src/mito2/src/schedule/priority.rs

```diff
@@ -0,0 +1,416 @@
+}
+
+/// Creates an unbounded channel.
+pub fn unbounded<T>() -> (Sender<T>, Receiver<T>) {
+    let (tx_low, rx_low) = async_channel::unbounded();
+    let (tx_high, rx_high) = async_channel::unbounded();
+    (Sender { tx_low, tx_high }, Receiver { rx_low, rx_high })
+}
+
+#[derive(Clone)]
+pub struct Sender<T> {
+    tx_low: async_channel::Sender<(Option<Instant>, T)>,
+    tx_high: async_channel::Sender<T>,
+}
+
+impl<T> Sender<T> {
+    /// Tries to send a high priority item, returns `Err` if the channel is full or closed.
+    pub fn try_send_high(&self, item: T) -> error::Result<()> {
+        self.tx_high.try_send(item).map_err(map_send_error)
+    }
+
+    /// Tries to send a low priority item with optional deadline, returns `Err` if the channel is full or closed.
+    pub fn try_send_low(&self, item: T, deadline: Option<Duration>) -> error::Result<()> {
+        self.tx_low
+            .try_send((deadline.map(|d| Instant::now().add(d)), item))
+            .map_err(map_send_error)
+    }
+
+    /// Sends a high priority item waiting for the channel to be ready.
+    /// Returns `Err` if the channel is closed.
+    pub async fn send_high(&self, item: T) -> error::Result<()> {
+        self.tx_high.send(item).await.map_err(map_send_error)
+    }
+
+    /// Sends a low priority with optional deadline waiting for the channel to be ready.
+    /// Returns `Err` if the channel is closed.
+    pub async fn send_low(&self, item: T, deadline: Option<Duration>) -> error::Result<()> {
+        self.tx_low
+            .send((deadline.map(|d| Instant::now().add(d)), item))
+            .await
+            .map_err(map_send_error)
+    }
+}
+
+fn map_send_error<E>(e: E) -> error::Error
+where
+    E: std::fmt::Debug,
+{
+    error::SendToChannelSnafu {
+        err_msg: format!("{:?}", e),
+    }
+    .build()
+}
+
+/// Channel receiver
+pub struct Receiver<T> {
+    rx_low: async_channel::Receiver<(Option<Instant>, T)>,
+    rx_high: async_channel::Receiver<T>,
+}
+
+impl<T> Clone for Receiver<T> {
+    fn clone(&self) -> Self {
+        Self {
+            rx_high: self.rx_high.clone(),
+            rx_low: self.rx_low.clone(),
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Converts the receiver into a async stream.
+    pub fn into_stream(self) -> impl Stream<Item = T> {
+        stream!({
+            let mut timer: Option<Pin<Box<Sleep>>> = None;
+            let mut pending: Option<T> = None;
+            loop {
+                select! {
+                    biased;
+                    Some(_) = maybe_timeout(&mut timer) => {
+                        if let Some(timed) = std::mem::take(&mut pending){
+                            yield timed
+                        }
+                    }
+
+                    Ok(high) = self.rx_high.recv() => {
+                        // also check if low channel has pending.
+                        if pending.is_none() && let Ok((deadline, low_item)) = self.rx_low.try_recv() {
```

Comment:
If we have many low-priority tasks exceed their deadline, it is still possible that we have no chance to execute the task with high prority.

Reply:
> If we have many low-priority tasks exceed their deadline, it is still possible that we have no chance to execute the task with high prority.

Maybe we should figure out a proper deadline.